core/iwasm/libraries/wasi-nn/src/wasi_nn_onnx.cpp

```diff
@@ -0,0 +1,818 @@
+    }
+
+    if (graph_index == -1) {
+        NN_ERR_PRINTF("Maximum number of graphs reached");
+        return runtime_error;
+    }
+
+    OrtStatus *status =
+        ctx->ort_api->CreateSession(ctx->env, name, ctx->session_options,
+                                    &ctx->graphs[graph_index].session);
+
+    if (status != nullptr) {
+        wasi_nn_error err = convert_ort_error_to_wasi_nn_error(ctx, status);
+        NN_ERR_PRINTF("Failed to create ONNX Runtime session from file: %s",
+                      name);
+        return err;
+    }
+
+    ctx->graphs[graph_index].is_initialized = true;
+    *g = graph_index;
+
+    NN_INFO_PRINTF("ONNX model loaded from file %s as graph %d", name,
+                   graph_index);
+    return success;
+}
+
+__attribute__((visibility("default"))) wasi_nn_error
+init_execution_context(void *onnx_ctx, graph g, graph_execution_context *ctx)
+{
+    OnnxRuntimeContext *ort_ctx = (OnnxRuntimeContext *)onnx_ctx;
+
+    if (g >= MAX_GRAPHS || !ort_ctx->graphs[g].is_initialized) {
+        NN_ERR_PRINTF("Invalid graph handle: %d", g);
+        return invalid_argument;
+    }
+
+    std::lock_guard<std::mutex> lock(ort_ctx->mutex);
+    int ctx_index = -1;
+    for (int i = 0; i < MAX_CONTEXTS; i++) {
+        if (!ort_ctx->exec_ctxs[i].is_initialized) {
+            ctx_index = i;
+            break;
+        }
+    }
+
+    if (ctx_index == -1) {
+        NN_ERR_PRINTF("Maximum number of execution contexts reached");
+        return runtime_error;
+    }
+
+    OnnxRuntimeExecCtx *exec_ctx = &ort_ctx->exec_ctxs[ctx_index];
+    exec_ctx->graph = &ort_ctx->graphs[g];
+
+    OrtStatus *status = ort_ctx->ort_api->CreateCpuMemoryInfo(
+        OrtArenaAllocator, OrtMemTypeDefault, &exec_ctx->memory_info);
+    if (status != nullptr) {
+        wasi_nn_error err = convert_ort_error_to_wasi_nn_error(ort_ctx, status);
+        NN_ERR_PRINTF("Failed to create CPU memory info");
+        return err;
+    }
+
+    size_t num_input_nodes;
+    status = ort_ctx->ort_api->SessionGetInputCount(exec_ctx->graph->session,
+                                                    &num_input_nodes);
+    if (status != nullptr) {
+        wasi_nn_error err = convert_ort_error_to_wasi_nn_error(ort_ctx, status);
+        ort_ctx->ort_api->ReleaseMemoryInfo(exec_ctx->memory_info);
+        NN_ERR_PRINTF("Failed to get input count");
+        return err;
+    }
+
+    for (size_t i = 0; i < num_input_nodes; i++) {
+        char *input_name;
+        status = ort_ctx->ort_api->SessionGetInputName(
+            exec_ctx->graph->session, i, ort_ctx->allocator, &input_name);
+        if (status != nullptr) {
+            wasi_nn_error err =
+                convert_ort_error_to_wasi_nn_error(ort_ctx, status);
+            ort_ctx->ort_api->ReleaseMemoryInfo(exec_ctx->memory_info);
+            NN_ERR_PRINTF("Failed to get input name");
+            return err;
+        }
+        exec_ctx->input_names.push_back(input_name);
```

Comment:
can push_back raise an exception?

Reply:
probably. i'm not sure how it should be handled.

maybe we can ignore the possibility for now.